torchtext/csrc/vocab.cpp

```diff
@@ -244,8 +268,61 @@ Vocab _load_vocab_from_file(const std::string &file_path,
 
     counter++;
     at::launch([&, file_path, num_lines, chunk_size, j, i, tokens_ptr]() {
-      parse_vocab_chunk(file_path, offsets[j], i,
-                        std::min(num_lines, i + chunk_size), tokens_ptr);
+      parse_vocab_file_chunk(file_path, offsets[j], i,
+                             std::min(num_lines, i + chunk_size), tokens_ptr);
+      std::lock_guard<std::mutex> lk(m);
+      counter--;
+      cv.notify_all();
+    });
+    chunk_tokens.push_back(tokens_ptr);
+    j++;
+  }
+
+  // block until all threads finish execution
+  std::unique_lock<std::mutex> lock(m);
+  cv.wait(lock, [&counter] { return counter == 0; });
+
+  IndexDict stoi;
+  StringList tokens;
+  std::tie(stoi, tokens) =
+      _concat_tokens(chunk_tokens, unk_token, min_freq, num_lines);
+
+  int64_t unk_index = stoi.find(unk_token)->second;
+
+  return Vocab(std::move(tokens), std::move(stoi), unk_token, unk_index);
+}
+
+Vocab _load_vocab_from_raw_text_file(const std::string &file_path,
+                                     const std::string &unk_token,
+                                     const int64_t min_freq,
+                                     const int64_t num_cpus, py::object fn) {
+  std::cerr << "[INFO] Reading file " << file_path << std::endl;
+
+  torch::jit::script::Module module = *torch::jit::as_module(fn);
```

Comment:
I think it's better to use the copy constructor for this or at a minimum verify that you can do concurrent access for the forward method and not cause any data races.

Reply:
Done